src/tty-ask-password-agent/tty-ask-password-agent.c

```diff
@@ -649,9 +673,167 @@ static int parse_argv(int argc, char *argv[]) {
+
+                for (ac = 0; ac < argc; ac++) {
+                        if (streq(argv[ac], "--console")) {
+                                argv[ac] = conarg;
+                                break;
+                        }
+                }
+
+                assert(ac < argc);
+
+                execv(SYSTEMD_TTY_ASK_PASSWORD_AGENT_BINARY_PATH, argv);
+                _exit(EXIT_FAILURE);
+        }
+        return 0;
+}
+
+static int ask_on_consoles(int argc, char *argv[]) {
+        _cleanup_set_free_ Set *pids = NULL;
+        _cleanup_strv_free_ char **consoles = NULL;
+        struct timespec timeout;
+        siginfo_t status = {};
+        sigset_t set;
+        Iterator i;
+        char **tty;
+        void *p;
+        pid_t pid;
+        int ret, signum;
+
+        ret = get_kernel_consoles(&consoles);
+        if (ret < 0)
+                return log_error_errno(ret, "Failed to determine devices of /dev/console: %m");
+
+        pids = set_new(NULL);
+        if (!pids)
+                return log_oom();
+
+        /* Start an agent on each console. */
+        STRV_FOREACH(tty, consoles) {
+                ret = ask_on_this_console(*tty, &pid, argc, argv);
+                if (ret < 0)
+                        return ret;
+
+                if (set_put(pids, PID_TO_PTR(pid)) < 0)
+                        return log_oom();
+        }
+
+        /* Wait for an agent to exit. */
+        for (;;) {
+                zero(status);
+
+                if (waitid(P_ALL, 0, &status, WEXITED) < 0) {
+                        if (errno == EINTR)
+                                continue;
+
+                        return log_error_errno(errno, "waitid() failed: %m");
+                }
+
+                set_remove(pids, PID_TO_PTR(status.si_pid));
+                break;
+        }
+
+        if (!is_clean_exit(status.si_code, status.si_status, NULL))
+                log_error("agent failed with: %d", status.si_status);
+
+        /*
+         * Request termination of the remaining processes as those
+         * are not required anymore.
+         */
+        SET_FOREACH(p, pids, i)
+                kill(PTR_TO_PID(p), SIGTERM);
+
+        /*
+         * Collect the processes which have go away.
+         */
+        assert_se(sigemptyset(&set) >= 0);
+        assert_se(sigaddset(&set, SIGCHLD) >= 0);
+        timespec_store(&timeout, 50 * USEC_PER_MSEC);
+
+        while (!set_isempty(pids)) {
+
+                zero(status);
+                ret = waitid(P_ALL, 0, &status, WEXITED|WNOHANG);
+                if (ret < 0 && errno == EINTR)
+                        continue;
+
+                if (!ret && status.si_pid > 0) {
+                        set_remove(pids, PID_TO_PTR(status.si_pid));
+                        continue;
+                }
+
+                signum = sigtimedwait(&set, NULL, &timeout);
+                if (signum < 0 && errno == EAGAIN)
+                        break;
+                assert(signum == SIGCHLD);
```

Comment:
It's probably my ignorance... Can you explain why you need both `waitid` and `sigtimedwait`?